File: plus_one/python_implementation/plus_one.py
#!/usr/bin/python3

"""
author: Andrew Burger

LeetCode problem:
O(n) Solution for adding one to a number whos digits are all a single
element in an array.
"""

from typing import List


class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:

        idx = len(digits) - 1
        solution = digits.copy()
        while idx >= 0:
            # solution[idx] != 9 case
            if solution[idx] != 9:
                solution[idx] = solution[idx] + 1
                break

            # solution[idx] == 9 case
            solution[idx] = 0

            # All elements are 9 i.e. [9,9,9,9]
            if idx - 1 < 0:
                solution.insert(0, 1)
                break

            idx -= 1

        return solution

def main():
    test1 = [9,9,9,9]
    test2 = [0]
    test3 = [1,2,3,4]
    tests = [test1, test2, test3]

    obj = Solution()

    for test in tests:
        solution = obj.plusOne(test)
        print("Test: %s" % (test))
        print("Solution: %s" % (solution))

if __name__ == "__main__":
    main()

File: plus_one/rust_implementation/plus_one.rs
fn plus_one(digits: Vec<i32>) -> Vec<i32> {

    let mut solution = digits.clone();
    let mut idx = digits.len() - 1;
    while idx >= 0 {

        if solution[idx] != 9 {
            solution[idx] += 1;
            break;
        }

        solution[idx] = 0;

        // special case all 9 i.e. [9,9,9,9]
        if idx as i32 - 1 < 0 {
            // insert 1 at beggining of vector
            solution.insert(0, 1);
            break;
        }

        idx -= 1;
    }

    solution
}

fn tests() {
    let mut test1 = vec![9,9,9,9];
    let mut test2 = vec![0];
    let mut test3 = vec![1,2,3];
    let mut tests = vec![test1,test2,test3];

    for test in tests.iter() {
        println!("Test: {:?}", test.to_vec());
        let solution = plus_one(test.to_vec());
        println!("Solution: {:?}", solution);
    }
}

fn main() {
    tests();
}
